File: Conversoes de Bases/Decimal e Octal/DecimalOctal.dart
import 'dart:io';
import 'dart:math';

const int LimiteMaximoDecimal = 2147483647;
const int LimiteMaximoOctal = 17777777777;

int decimalParaOctal(int decimal) {
  int resto, potencia = 1, octal = 0;
  while (decimal > 0) {
    resto = decimal % 8;
    octal += resto * potencia;
    decimal ~/= 8;
    potencia *= 10;
  }
  return octal;
}

int octalParaDecimal(int octal) {
  int potencia = 0, digito, decimal = 0;
  while (octal > 0) {
    digito = octal % 10;
    decimal += digito * pow(8, potencia).toInt();
    octal ~/= 10;
    potencia++;
  }
  return decimal;
}

bool isValidOctal(int octal) {
  while (octal > 0) {
    if (octal % 10 > 7) {
      return false;
    }
    octal ~/= 10;
  }
  return true;
}

void main() {
  int opcao;
  int decimal;
  int octal;
  do {
    print('=========================');
    print(' CONVERSOR DECIMAL-OCTAL       ');
    print('==========================');
    print('');
    print('1 - Converter decimal para octal');
    print('2 - Converter octal para decimal');
    print('0 - Sair');
    print('');
    stdout.write('Digite a opcao desejada: ');
    opcao = int.parse(stdin.readLineSync()!);
    switch (opcao) {
      case 1:
        print('');
        stdout.write('Digite um numero decimal (ate $LimiteMaximoDecimal): ');
        decimal = int.parse(stdin.readLineSync()!);
        if (decimal > LimiteMaximoDecimal) {
          print('Valor decimal fora do limite permitido.');
        } else {
          print('Valor em octal: ${decimalParaOctal(decimal)}');
        }
        break;
      case 2:
        print('');
        stdout.write('Digite um numero octal (ate $LimiteMaximoOctal): ');
        octal = int.parse(stdin.readLineSync()!);
        if (!isValidOctal(octal)) {
          print('Valor octal invalido.');
        } else if (octal > LimiteMaximoOctal) {
          print('Valor octal fora do limite permitido.');
        } else {
          print('Valor em decimal: ${octalParaDecimal(octal)}');
        }
        break;
      case 0:
        break;
      default:
        print('Opcao invalida. Digite novamente.');
    }
    print('');
  } while (opcao != 0);
}

File: Conversoes de Bases/Decimal e Octal/DecimalOctal.go
package main

import (
	"fmt"
	"math"
)

const (
	LimiteMaximoDecimal uint32 = 2147483647
	LimiteMaximoOctal   uint32 = 017777777777
)

func DecimalParaOctal(decimal uint32) uint32 {
	var resto, potencia, octal uint32
	octal = 0
	potencia = 1
	for decimal > 0 {
		resto = decimal % 8
		octal += resto * potencia
		decimal /= 8
		potencia *= 10
	}
	return octal
}

func OctalParaDecimal(octal uint32) uint32 {
	var potencia, digito, decimal uint32
	decimal = 0
	potencia = 0
	for octal > 0 {
		digito = octal % 10
		decimal += digito * uint32(math.Pow(8, float64(potencia)))
		octal /= 10
		potencia++
	}
	return decimal
}

func IsValidOctal(octal uint32) bool {
	for octal > 0 {
		if octal%10 > 7 {
			return false
		}
		octal /= 10
	}
	return true
}

func main() {
	var opcao int
	var decimal, octal uint32
	for {
		fmt.Println("=========================")
		fmt.Println(" CONVERSOR DECIMAL-OCTAL ")
		fmt.Println("=========================")
		fmt.Println()
		fmt.Println("1 - Converter decimal para octal")
		fmt.Println("2 - Converter octal para decimal")
		fmt.Println("0 - Sair")
		fmt.Println()
		fmt.Print("Digite a opcao desejada: ")
		fmt.Scanln(&opcao)
		switch opcao {
		case 1:
			fmt.Println()
			fmt.Printf("Digite um numero decimal (ate %d): ", LimiteMaximoDecimal)
			fmt.Scanln(&decimal)
			if decimal > LimiteMaximoDecimal {
				fmt.Println("Valor decimal fora do limite permitido.")
			} else {
				fmt.Println("Valor em octal:", DecimalParaOctal(decimal))
			}
		case 2:
			fmt.Println()
			fmt.Printf("Digite um numero octal (ate %o): ", LimiteMaximoOctal)
			fmt.Scanln(&octal)
			if !IsValidOctal(octal) {
				fmt.Println("Valor octal invalido.")
			} else if octal > LimiteMaximoOctal {
				fmt.Println("Valor octal fora do limite permitido.")
			} else {
				fmt.Println("Valor em decimal:", OctalParaDecimal(octal))
			}
		}
		fmt.Println()
		if opcao == 0 {
			break
		}
	}
}
